@xen-orchestra/proxy/src/main/index.js

```diff
@@ -1,13 +1,43 @@
-import { camelCase } from 'lodash'
+import camelCase from 'lodash/camelCase'
 
 import mixins from './mixins'
 
+const { defineProperties, defineProperty, keys } = Object
+const noop = Function.prototype
+
+const MIXIN_CYCLIC_DESCRIPTOR = {
+  configurable: true,
+  get() {
+    throw new Error('cyclic dependency')
+  },
+}
+
 export default class App {
   constructor(opts) {
-    Object.keys(mixins).forEach(mixin => {
-      Object.defineProperty(this, camelCase(mixin), {
-        value: new mixins[mixin](this, opts),
-      })
+    // add lazy property for each of the mixin, this allow mixins to depend on
+    // one another without any special ordering
+    const descriptors = {}
+    keys(mixins).forEach(name => {
+      const Mixin = mixins[name]
+      name = camelCase(name)
+
+      descriptors[name] = {
+        configurable: true,
+        get: () => {
+          defineProperty(this, name, MIXIN_CYCLIC_DESCRIPTOR)
+          const instance = new Mixin(this, opts)
+          defineProperty(this, name, {
+            value: instance,
+          })
+          return instance
+        },
+      }
+    })
+    defineProperties(this, descriptors)
+
+    // ensure all mixins are instanciated
+    keys(descriptors).forEach(name => {
```

Comment:
How it ensure that mixins are instanciated?

Reply:
It accesses all the properties to trigger the getters.